pkg/volume/csi/csi_client_test.go

```diff
@@ -338,3 +388,100 @@ func TestClientNodeUnstageVolume(t *testing.T) {
 		}
 	}
 }
+
+func TestGetNodeInfoWithFallback(t *testing.T) {
+	var errRandom = errors.New("some other error")
+	var errUnimplemented = status.Error(codes.Unimplemented, "unknown method someRandomMethod")
+
+	testCases := map[string]struct {
+		nodeGetInfoErr          error
+		nodeGetIdErr            error
+		expectedErr             error
+		expectedGetNodeInfoCall bool
+		expectedGetNodeIdCall   bool
+	}{
+		"no error on neither method": {
+			nil,
+			nil,
+			nil,
+			true,
+			false,
+		},
+		"unimplemented NodeGetInfo": {
+			errUnimplemented,
+			nil,
+			nil,
+			true,
+			true,
+		},
+		"other error for NodeGetInfo": {
+			errRandom,
+			nil,
+			errRandom,
+			true,
+			false,
+		},
+		"unimplemented NodeGetInfo and error on NodeGetId": {
+			errUnimplemented,
+			errRandom,
+			errRandom,
+			true,
+			true,
+		},
+	}
+
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			ctx := context.TODO()
+			ig := &nodeInfoGetter{
+				nodeGetInfoErr: tc.nodeGetInfoErr,
+				nodeGetIdErr:   tc.nodeGetIdErr,
+			}
+			_, _, _, err := getNodeInfoWithFallback(ctx, ig)
+			if err != tc.expectedErr {
+				t.Errorf("Expected error %v, got: %v", tc.expectedErr, err)
+			}
+			checkCall(t, ig.nodeGetInfoCallCount, tc.expectedGetNodeInfoCall, "NodeGetInfo")
```

Comment:
Instead of injecting/checking for errors like this, could you do it in the same way the rest of the code is written?
For example with the fake client here
```
client.(*fakeCsiDriverClient).nodeClient.SetNextError(tc.err)
client.(*fakeCsiDriverClient).nodeClient.SetNodeGetIdResp(&csipb.NodeGetIdResponse{
	NodeId: tc.expectedNodeID,
})
```

Reply:
I am not sure what you actually mean here. Is it something along the lines like in [this experiment](https://github.com/hoegaarden/kubernetes/commit/4316e05f3ba2477dbeb8e66f72428ac006b39695)?

Reply:
yes, I think you understood correctly. that looks better to me

Reply:
We agree that we could use the existing test infrastructure, as shown in the experiment. One concern we have doing this is that it seems to be unnecessarily complicated: Our code (`getNodeInfoWithFallback`) calls two methods (`NodeGetInfo` & `NodeGetId`) on an object that implements the `nodeInfoGetter` interface - that’s all it needs for the test. Introducing the `fakeCsiDriverClient` which uses a nested `fakeNodeClient` and stubs out some methods on that nested fake seems to be way more complicated than it needs to be. The test then needs to care about nested implementation that it shouldn’t need to care about.

A larger problem with this whole nested fakes thing is that the tests for the `CsiDriverClient` (at least for `NodeGetInfo` and `NodeGetId`, probably the other methods) actually **just test the fakes** (`fakeCsiDriverClient` and `fakeNodeClient`) and **never actually exercise production code**. We have verified that the production code is untested by changing the implementation of `NodeGetInfo` on `CsiDriverClient` to never return an error (change `err` to `nil` [here](https://github.com/kubernetes/kubernetes/blob/807eb6b24b953ff3f59c8049eff7e0edbc279e18/pkg/volume/csi/csi_client.go#L99)). If we make this change, all the tests still pass, even the test that [looks like it was intended to catch this problem](https://github.com/kubernetes/kubernetes/blob/807eb6b24b953ff3f59c8049eff7e0edbc279e18/pkg/volume/csi/csi_client_test.go#L187). So the existing testing strategy here is not helping to test the production code, and the fact that this is not obvious means the test code is hard to reason about, so we’re reluctant to continue that pattern with our new test that this discussion section is about.

We’d like to raise and track this issue we found with the test strategy properly. What is the preferred way to do so? GH issue? Discuss in some SIG meeting? Anything else?

Also, with all the details outlined above, do you still want us to change our test setup?

Reply:
@flangewad thanks for the detailed explaination. I think a Github issue would be appropriate, make sure to cc me and whoever else you think might be able to help (the original authors of the test maybe). If I understand correctly this is actually a significant issue of many existing tests not actually testing production behavior at all? This may be worth bringing up at a Kubernetes-CSI implementation meeting, the ones that happen every mon-wed-fri.

Is this issue also present in the current iteration of the test in this PR? If not, then lets keep it this way and get it merged. If so, can this PR wait until the more general testing issue is solved?

Reply:
@davidz627 Great, we will investigate to see the extent of the problem and raise an issue.

In terms of our code, the test that we added does not have this problem because it doesn't use the problematic fakes. So it sounds like you'd be happy to merge?🤞

Reply:
@davidz627 new issue [here](https://github.com/kubernetes/kubernetes/issues/69219)